File: frontend/src/services/incidents.ts
// src/services/incidents.ts

import api from './api';
import { apiGet, apiPost, apiPatch, apiDelete } from './api';
import type { ApiResponse } from '../types/api';
import type { Incident, IncidentCreate, IncidentUpdate, IncidentListMeta, IncidentStats } from '../types/incident';

export interface ListIncidentsParams {
  page?:          number;
  page_size?:     number;
  incident_id?:   string;
  category?:      string;
  severity?:      string;
  status?:        string;
  business_unit?: string;
  search?:        string;
}

export interface IncidentListResponse {
  items: Incident[];
  meta:  IncidentListMeta;
}

export async function listIncidents(params: ListIncidentsParams = {}): Promise<IncidentListResponse> {
  const q = new URLSearchParams();
  if (params.page)          q.set('page',          String(params.page));
  if (params.page_size)     q.set('page_size',     String(params.page_size));
  if (params.incident_id)   q.set('incident_id',   params.incident_id);
  if (params.category)      q.set('category',      params.category);
  if (params.severity)      q.set('severity',      params.severity);
  if (params.status)        q.set('status',        params.status);
  if (params.business_unit) q.set('business_unit', params.business_unit);
  if (params.search)        q.set('search',        params.search);

  const qs  = q.toString();
  const url = `/api/v1/incidents${qs ? `?${qs}` : ''}`;
  const envelope = await api.get<ApiResponse<Incident[]>>(url);
  if (envelope.data.error) throw new Error(envelope.data.error);

  return {
    items: envelope.data.data ?? [],
    meta:  envelope.data.meta as unknown as IncidentListMeta,
  };
}

export async function createIncident(payload: IncidentCreate): Promise<Incident> {
  return apiPost<Incident>('/api/v1/incidents', payload);
}

export async function updateIncident(id: string, payload: IncidentUpdate): Promise<Incident> {
  return apiPatch<Incident>(`/api/v1/incidents/${id}`, payload);
}

export async function deleteIncident(id: string): Promise<void> {
  await apiDelete(`/api/v1/incidents/${id}`);
}

export async function getIncidentStats(): Promise<IncidentStats> {
  return apiGet<IncidentStats>('/api/v1/incidents/stats');
}

export async function generateAIImpact(id: string, force = false): Promise<{ mode: string; text: string }> {
  return apiPost(`/api/v1/incidents/${id}/ai/impact`, { force });
}

export async function generateAIActions(id: string, force = false): Promise<{ mode: string; text: string }> {
  return apiPost(`/api/v1/incidents/${id}/ai/actions`, { force });
}

export async function suggestCategory(description: string): Promise<string> {
  const res = await apiPost<{ suggestion: string }>('/api/v1/incidents/ai/suggest-category', { description });
  return res.suggestion;
}

export async function suggestSeverity(description: string): Promise<string> {
  const res = await apiPost<{ suggestion: string }>('/api/v1/incidents/ai/suggest-severity', { description });
  return res.suggestion;
}

File: backend/app/api/v1/routes/incidents.py
# app/api/v1/routes/incidents.py

from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_db, get_active_tenant, require_module, require_permission
from app.schemas.incident import (
    IncidentCreate,
    IncidentUpdate,
    AIIncidentRequest,
    AIIncidentSuggestRequest,
)
from app.services import incident as incident_service
from app.services import ai_incident as ai_incident_service

router = APIRouter(prefix="/incidents", tags=["incidents"])


@router.get("/stats")
async def get_stats(
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_module("incident")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    stats = await incident_service.get_stats(db, tenant_id)
    return {"data": stats, "error": None, "meta": {}}


@router.get("")
async def list_incidents(
    page: int               = Query(1, ge=1),
    page_size: int          = Query(10, ge=1, le=200),
    incident_id: str | None = Query(None),
    category: str | None    = Query(None),
    severity: str | None    = Query(None),
    status: str | None      = Query(None),
    business_unit: str | None = Query(None),
    search: str | None      = Query(None),
    db: AsyncSession        = Depends(get_db),
    claims: dict            = Depends(require_module("incident")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    result = await incident_service.list_incidents(
        db, tenant_id, page, page_size,
        incident_id, category, severity, status, business_unit, search,
    )
    return {
        "data": result.items,
        "error": None,
        "meta": {
            "total":     result.total,
            "page":      result.page,
            "page_size": result.page_size,
        },
    }


@router.post("")
async def create_incident(
    payload: IncidentCreate,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_permission("manage_incidents")),
    _:       dict    = Depends(require_module("incident")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    inc = await incident_service.create_incident(
        db, tenant_id, payload, claims["email"]
    )
    return {"data": inc, "error": None, "meta": {}}


@router.patch("/{incident_id}")
async def update_incident(
    incident_id: str,
    payload: IncidentUpdate,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_permission("manage_incidents")),
    _:       dict    = Depends(require_module("incident")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    inc = await incident_service.update_incident(
        db, tenant_id, incident_id, payload, claims["email"]
    )
    return {"data": inc, "error": None, "meta": {}}


@router.delete("/{incident_id}")
async def delete_incident(
    incident_id: str,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_permission("manage_incidents")),
    _:       dict    = Depends(require_module("incident")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    await incident_service.delete_incident(
        db, tenant_id, incident_id, claims["email"]
    )
    return {
        "data": {"message": f"Incident {incident_id} moved to recycle bin"},
        "error": None,
        "meta": {},
    }


@router.post("/{incident_id}/ai/impact")
async def generate_impact(
    incident_id: str,
    payload: AIIncidentRequest,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_permission("generate_ai")),
    _:       dict    = Depends(require_module("incident")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    result = await ai_incident_service.generate_impact(
        db, tenant_id, incident_id, payload.force
    )
    return {"data": result, "error": None, "meta": {}}


@router.post("/{incident_id}/ai/actions")
async def generate_actions(
    incident_id: str,
    payload: AIIncidentRequest,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_permission("generate_ai")),
    _:       dict    = Depends(require_module("incident")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    result = await ai_incident_service.generate_actions(
        db, tenant_id, incident_id, payload.force
    )
    return {"data": result, "error": None, "meta": {}}


@router.post("/ai/suggest-category")
async def suggest_category(
    payload: AIIncidentSuggestRequest,
    db:      AsyncSession = Depends(get_db),
    claims:  dict         = Depends(require_permission("generate_ai")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    result = await ai_incident_service.suggest_category(db, tenant_id, payload.description)
    return {"data": result, "error": None, "meta": {}}


@router.post("/ai/suggest-severity")
async def suggest_severity(
    payload: AIIncidentSuggestRequest,
    db:      AsyncSession = Depends(get_db),
    claims:  dict         = Depends(require_permission("generate_ai")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    result = await ai_incident_service.suggest_severity(db, tenant_id, payload.description)
    return {"data": result, "error": None, "meta": {}}